app/plugin_install.go

```diff
@@ -369,3 +307,88 @@ func (a *App) getBundleStorePath(id string) string {
 func (a *App) getSignatureStorePath(id string, index int) string {
 	return filepath.Join(fileStorePluginFolder, fmt.Sprintf("%s.%d.sig", id, index))
 }
+
+func extractPlugin(pluginFile io.ReadSeeker, extractDir string) (*model.Manifest, string, error) {
+	pluginFile.Seek(0, 0)
+	if err := utils.ExtractTarGz(pluginFile, extractDir); err != nil {
+		return nil, extractDir, errors.New("Encountered error extracting plugin")
+	}
+
+	dir, err := ioutil.ReadDir(extractDir)
+	if err != nil {
+		return nil, extractDir, errors.Errorf("Can't read directory %s", extractDir)
+	}
+
+	if len(dir) == 1 && dir[0].IsDir() {
+		extractDir = filepath.Join(extractDir, dir[0].Name())
+	}
+
+	manifest, _, err := model.FindManifest(extractDir)
+
+	if err != nil {
+		return nil, extractDir, errors.Wrapf(err, "Can't find manifest in directory %s", extractDir)
+	}
+
+	if !plugin.IsValidId(manifest.Id) {
+		return nil, extractDir, errors.Errorf("Plugin Id must be at least %d characters, at most %d characters and match %d.", plugin.MinIdLength, plugin.MaxIdLength, http.StatusBadRequest)
```

Comment:
Why are you appending `http.StatusBadRequest` here?
```suggestion
		return nil, extractDir, errors.Errorf("Plugin Id must be at least %d characters, at most %d characters and match %d.", plugin.MinIdLength, plugin.MaxIdLength, plugin.ValidIdRegex})
```

Reply:
Restored the original app error codes here.